source/dofs/dof_handler.cc

```diff
@@ -1002,8 +1000,8 @@ void DoFHandler<dim,spacedim>::distribute_dofs (const FiniteElement<dim,spacedim
 
   // Only recreate the FECollection if we don't already store
   // the exact same FiniteElement object.
-  if (fe_collection == nullptr || &((*fe_collection)[0]) != &ff)
-    fe_collection = std_cxx14::make_unique<hp::FECollection<dim, spacedim>>(ff);
+  if  (fe_collection.size() == 0 || &(fe_collection[0]) != &ff)
```

Comment:
I do not believe that this is correct since `FECollection` will clone finite element objects.

Reply:
The idea is that when we are trying to (re-)`distribute_dofs` in a `DoFHandler` object with the `FiniteElement` obtained from calling `get_fe()` on the same object we are not creating the `FiniteElement` anew. This is needed and tested in some test that @luca-heltai came up with. If you want I can try to find it tomorrow.

You are right that calling `distribute_dofs` multiple times with the same `FiniteElement` doesn't trigger this if statement.

Reply:
You're right; I did not think of that use case. My bad.

Reply:
This is not working as expected. Apparently we do not test this throughly. I'll make a PR with a couple of failing tests.

Reply:
(in particular, calling `distribute_dofs` multiple times **does** trigger this. The address of the stored fe_collection will not be the same of the finite element passed to this guy, since in fact fe_collection *does clone* the fe internally.

Reply:
see #6325.